docs/ErrorHandling.md

```diff
@@ -0,0 +1,41 @@
+# Handling errors in MLIR
+
+Three are two different kinds of errors: errors that comes from user inputs, and compiler errors. We should provide meaningful user feedback for user input errors, and are typically handled using `emitError` functions. Compiler errors should be reported using `asserts` or `llvm_unreachable` calls.
+
```

Comment:
It seems that the different error report method is chosen based on the calling context. EmitError is for functions with LogicalResult return. Not clear about how to choose between asserts and llvm_unreachable.
For semantics, how should we systematically distinguish different reasons, such as incorrect input, compiler inconsistence, unimplemented case or unexpected case.

Reply:
@chentong319 added the requested update